x-pack/legacy/plugins/apm/public/components/shared/Links/url_helpers.ts

```diff
@@ -3,20 +3,17 @@
  * or more contributor license agreements. Licensed under the Elastic License;
  * you may not use this file except in compliance with the Elastic License.
  */
-
-import qs from 'querystring';
 import { LocalUIFilterName } from '../../../../server/lib/ui_filters/local_ui_filters/config';
+import { url } from '../../../../../../../../src/plugins/kibana_utils/public';
```

Comment:
Any reason not to use:

```
import { parse } from 'query-string';
```

Reply:
@sqren we can, but why should we do this if we have one common method for doing this

Reply:
It just seems a bit confusing that we have our own abstraction on top of the query-string lib. If we used the query-string's lib `parse` method directly it's very easy to find docs: https://github.com/sindresorhus/query-string#parsestring-options

If I wanted to find docs for `url` method, where would I find that (Tried Googling `kibana_utils url` which obviously doesn't give much).

Reply:
@sqren 

it's a reason why we need to introduce one common util and use it across the whole Kibana. Let me explain why I think so. Please scroll down to :

```
export function fromQuery(query: Record<string, any>) {
  return url.stringifyUrlQuery(query, (value: string) =>
    encodeURIComponent(value).replace(/%3A/g, ':')
  );
}
```

Could you please  explain why `apm` do replace only  `(/%3A/g, ':')` but in almost all of our other places we do: 

```
  return encodeURIComponent(val)
    .replace(/%40/gi, '@')
    .replace(/%3A/gi, ':')
    .replace(/%24/g, '$')
    .replace(/%2C/gi, ',')
    .replace(/%3B/gi, ';')
    .replace(/%20/g, pctEncodeSpaces ? '%20' : '+');
```
Probably we should try to find how the same logic was implemented in other places and then ask google and implement your own code

Reply:
@sqren we're actually trying to figure out why every solution has a slightly different implementation of this logic. It's very confusing to see different implantation with varying character escaping.

Putting documentation aside ([we're working on this now](https://github.com/elastic/kibana/issues/56881)), is there a reason the logic above won't suit your needs?

Reply:
>  is there a reason the logic above won't suit your needs?

Without documentation about what that API does, I can't answer that questions for sure :) I think there are valid reasons for different apps to have different needs. Apps in Angular will have different issues wrt url encoding/decoding than apps using react-router. Similarly apps that uses rison will have different needs from those that don't.

If we can have one solution that fits all I think that is great.

Reply:
> Could you please explain why apm do replace only (/%3A/g, ':') 

Ideally simply using `qs.stringify(query)` would suffice. Unfortunately the angular router (which affects all apps in Kibana, even non-angular ones) will rewrite the url so urlencoded colons (`%3A`) are replaced with their decoded value `:`. To avoid having duplicate entires in the location history we encode links with this limitation in mind. In NP when angular is removed from core I expect that we can simply use `qs.stringify(query)` without needing any custom encoding.

> but in almost all of our other places we do:

I'm not sure why more custom encoding is needed. But I'm sure people have valid reasons (different apps, different requirements).

Reply:
To be clear, I think the end goal should be for us to reduce the need for custom encoding, and I'm worried that introducing the suggested abstraction will enforce the status quo. 
The APM plugin will be "custom encoding free" in NP - wouldn't it be better to try to achieve the same for the rest of the platform?

Reply:
I think I'm in agreement with @sqren here. It feels to me like everybody should simply be using the `query-string` npm lib, and only using a common util if they need/want to.

Especially if it is in fact possible for `encode_query_component` to go away in NP (I am making this assumption, but haven't tested to verify), at which point this wrapper in `kibana_utils` basically won't be doing anything.

Instead of wrapping `query-string` inside of `stringifyUrlQuery` and `parseUrlQuery`, what if we instead did the following:
1. Get rid of `parseUrlQuery` because it is doing nothing.
2. Export `encode_query_component` for folks who _want_ to use this together with `stringify`, or alternatively: create a smaller util that can do the lodash `transform` part for you

This has a few advantages:
- Everything becomes opt-in: default to using the well-established NPM lib, and use the `kibana_utils` helpers alongside it if you need to. 
- Apps have full control over how they want to handle their encoding
- It's not creating more abstraction than we need, and is less surface area for kibana_utils / app arch to maintain